### PR TITLE
feat(provider/cf): add model and cache keys

### PR DESCRIPTION
**Pull request by jkschneider**
Adds Cloud Foundry implementation of Spinnaker model objects.

Previous PRs in this sequence: #2838, #2828.

### CODE REVIEW

File: clouddriver-cloudfoundry/clouddriver-cloudfoundry.gradle [hunk 1]
```diff
@@ -10,7 +14,9 @@ dependencies {
   compile spinnaker.dependency('lombok')
 
   spinnaker.group('retrofitDefault')
-  spinnaker.group('test')
+
+  testCompile('org.junit.jupiter:junit-jupiter-api:5.2.0')
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/clouddriver-cloudfoundry.gradle]: Is there a group similar to `spinnaker.group('test')` that we can create for this?

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/Cluster.java [hunk 2]
```diff
@@ -69,7 +69,7 @@ default Moniker getMoniker() {
    */
   @Empty
   @JsonSerialize(nullsUsing = NullCollectionSerializer.class)
-  Set<ServerGroup> getServerGroups();
+  Set<? extends ServerGroup> getServerGroups();
```
Comment (jkschneider) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/Cluster.java]: Allows us to use Lombok `@Getter` on model implementations while holding a more specific implementation of `ServerGroup`, e.g. `CloudFoundryServerGroup`. Much less casting when using the model object later.

Due to Groovy's relaxed type system, this likely never came up before.

**Comment by jkschneider** [comment]
cc / @ethanfrogers

File: clouddriver-cloudfoundry/clouddriver-cloudfoundry.gradle (hunk 1 above)
Comment (lwander) [reply clouddriver-cloudfoundry/clouddriver-cloudfoundry.gradle]: Not that I know of, but a new one would be welcome: github.com/spinnaker/spinnaker-dependencies

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/cache/Keys.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.Map;
+
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.CloudFoundryCloudProvider.ID;
+import static com.netflix.spinnaker.clouddriver.cloudfoundry.cache.Keys.Namespace.*;
+
+@Component("CloudFoundryInfraKeys")
+public class Keys implements KeyParser {
+
+  @Nullable
+  public static Map<String, String> parse(String key) {
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/cache/Keys.java]: I've found it's been helpful to keep this as `Optional` for later usage, but up to you style wise

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  private final Set<CloudFoundryServerGroup> mappedApps;
+
+  @JsonProperty
+  public String getName() {
+    return host + "." + domain + "." + getName() + (port == null ? "" : "-" + port) + (path == null ? "" : "/" + path);
+  }
+
+  @Override
+  public Moniker getMoniker() {
+    return EMPTY_MONIKER;
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java]: I would recommend setting this correctly unless there is an explicit need for keeping this empty (or you plan to implement this later).

Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java]: Yes, there is an explicit need to keep it empty. Cloud Foundry load balancers are inseparable from public routes in a way that a moniker doesn't describe well.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/cache/Keys.java (hunk 3 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/cache/Keys.java]: Thanks. I'm comfortable with the way it is if you are.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java (hunk 4 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java]: In other words, it isn't possible to separate a CF load balancer into cluster, stack, detail, etc.

Comment (lwander) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java]: What about the 'app' field?

Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/model/CloudFoundryLoadBalancer.java]: There is also no way to infer an app from a CF load balancer.

**Pull request merged**